packages/mobile/src/geth/saga.ts

```diff
@@ -192,3 +212,167 @@ export function* gethSagaIfNecessary() {
+    const keys = yield call(generateKeys, mnemonic, undefined, undefined, bip39)
+    const privateKey = keys.privateKey
+    if (!privateKey) {
+      throw new Error('Failed to convert mnemonic to hex')
+    }
+
+    const accountAddress = yield call(assignAccountFromPrivateKey, privateKey)
+    if (!accountAddress) {
+      throw new Error('Failed to assign account from private key')
+    }
+
+    yield call(setKey, 'mnemonic', mnemonic)
+
+    return accountAddress
+  } catch (error) {
+    // Capturing error in sentry for now as we debug backup key issue
+    Sentry.captureException(error)
+    Logger.error(TAG + '@getOrCreateAccount', 'Error creating account', error)
+    throw new Error(ErrorMessages.ACCOUNT_SETUP_FAILED)
+  }
+}
+
+export function* assignAccountFromPrivateKey(privateKey: string) {
+  try {
+    const pincode = yield call(getPincode, false)
+    if (!pincode) {
+      Logger.error(TAG + '@assignAccountFromPrivateKey', 'Got falsy pin')
+      throw Error('Cannot create account without having the pin set')
+    }
+
+    // NOTE: currentAccount can be null
+    const currentAccount = yield select(currentAccountSelector)
+    const account = yield call(addAccountToGethKeystore, privateKey, currentAccount, pincode)
+    Logger.debug(TAG + '@assignAccountFromPrivateKey', `Added to wallet: ${account}`)
+
+    yield put(setAccount(account))
+    yield put(setAccountCreationTime())
+    yield call(assignDataKeyFromPrivateKey, privateKey)
+    return account
+  } catch (e) {
+    Logger.error(TAG + '@assignAccountFromPrivateKey', 'Error assigning account', e)
+    throw e
+  }
+}
+
+function* assignDataKeyFromPrivateKey(privateKey: string) {
+  const privateCEK = deriveCEK(privateKey).toString('hex')
+  yield put(setPrivateCommentKey(privateCEK))
+}
+
+export function* getAccount() {
+  while (true) {
+    const account = yield select(currentAccountSelector)
+    if (account) {
+      return account
+    }
+
+    const action = yield take([Actions.SET_ACCOUNT, REHYDRATE])
+    if (action.type === REHYDRATE) {
+      // Wait for rehydrate and select the state again
+      continue
+    }
+    if (action.address) {
+      // account exists
+      return action.address
+    }
+  }
+}
+
+// Stores account and private key in geth keystore
+export function* addAccountToGethKeystore(key: string, currentAccount: string, pincode: string) {
+  Logger.debug(TAG + '@addAccountToGethKeystore', `using key ${key} for account ${currentAccount}`)
+
+  let account: string
+  try {
+    yield call(waitForGethConnectivity)
+    account = yield call(gethWallet.addAccount, key, pincode)
+    Logger.debug(
+      TAG + '@addAccountToGethKeystore',
+      `Successfully imported raw key for account ${account}`
+    )
+    yield put(setAccountInGethKeystore(account))
+  } catch (e) {
+    Logger.error(TAG + '@addAccountToGethKeystore', 'Failed to import raw key', e)
+    if (e.toString().includes('account already exists')) {
```

Comment:
think this is what you're talking about in terms of error parsing. i'm not really sure there is any other way to check this save to look for an enum in the blockchain codebase, but if you think of one it would definitely be nicer